core/dbt/exceptions.py

```diff
@@ -1803,6 +1803,16 @@ def get_message(self) -> str:
 
 
 # jinja exceptions
+class ColumnTypeMissingError(CompilationError):
+    def __init__(self, column_names: List):
+        self.column_names = column_names
+        super().__init__(msg=self.get_message())
+
+    def get_message(self) -> str:
+        msg = f"Columns with `data_type` Blank/Null not allowed on contracted models. Columns Blank/Null: {self.column_names}"
+        return msg
```

Comment:
@dbeatty10 - would love your input on this message.  

This yaml

```yaml
  - name: my_model
    columns:
      - name: id
        data_type: integer
      - name: color
      - name: date_day
        data_type: date
```

gives this output

```zsh
19:54:32  Finished running 2 view models, 1 table model in 0 hours 0 minutes and 0.59 seconds (0.59s).
19:54:32
19:54:32  Completed with 1 error and 0 warnings:
19:54:32
19:54:32  Compilation Error in model my_model (models/contracted_models/my_model.sql)
19:54:32    Columns with `data_type` Blank/Null not allowed on contracted models. Columns Blank/Null: ['color']
19:54:32
19:54:32    > in macro default__get_empty_schema_sql (macros/adapters/columns.sql)
19:54:32    > called by macro get_empty_schema_sql (macros/adapters/columns.sql)
19:54:32    > called by macro assert_columns_equivalent (macros/materializations/models/table/columns_spec_ddl.sql)
19:54:32    > called by macro default__get_assert_columns_equivalent (macros/materializations/models/table/columns_spec_ddl.sql)
19:54:32    > called by macro get_assert_columns_equivalent (macros/materializations/models/table/columns_spec_ddl.sql)
19:54:32    > called by macro default__create_view_as (macros/materializations/models/view/create_view_as.sql)
19:54:32    > called by macro create_view_as (macros/materializations/models/view/create_view_as.sql)
19:54:32    > called by macro default__get_create_view_as_sql (macros/materializations/models/view/create_view_as.sql)
19:54:32    > called by macro get_create_view_as_sql (macros/materializations/models/view/create_view_as.sql)
19:54:32    > called by macro statement (macros/etc/statement.sql)
19:54:32    > called by macro materialization_view_default (macros/materializations/models/view/view.sql)
19:54:32    > called by model my_model (models/contracted_models/my_model.sql)
19:54:32
19:54:32  Done. PASS=1 WARN=0 ERROR=1 SKIP=1 TOTAL=3
```

Reply:
These have to be defined in YAML rather than model config in SQL/Python, right?

If so, how 'bout this?

> Contracted models require `data_type` to be defined for each column. Please ensure that the column `name` and `data_type` are defined within the YAML configuration for the ['color'] column(s).

It's a bit verbose, but hopefully the problem and solution are both clear.